PERF: releasing the GIL, #8882

jreback opened the pull request:
closes #8882
closes #10139 

This is now implemented for all groupbys with the factorization part as well (which is actually a big part fo the time)

```
In [2]: N = 1000000

In [3]: ngroups = 1000

In [4]: np.random.seed(1234)

In [5]: df = DataFrame({'key' : np.random.randint(0,ngroups,size=N),
   ...:                 'data' : np.random.randn(N) })
```

```
def f():
    df.groupby('key')['data'].sum()

# run consecutivily
def g2():
    for i in range(2):
        f()
def g4():
    for i in range(4):
        f()
def g8():
    for i in range(8):
        f()

# run in parallel
@test_parallel(num_threads=2)
def pg2():
    f()

@test_parallel(num_threads=4)
def pg4():
    f()

@test_parallel(num_threads=8)
def pg8():
    f()
```

So seeing a nice curve as far as addtl cores are used (as compared to some of my previous posts).

```
In [19]: df
Out[19]: 
   after  before   speedup
2   27.3    52.1  1.908425
4   44.8   105.0  2.343750
8   83.2   213.0  2.560096
```

jreback committed:
PERF: releasing the GIL, #8882

jreback committed:
PERF: vbenches for #8882, releasing the GIL

mrocklin commented:
What was the deal with `NA` and `nan`.  These were Python objects?

jreback commented:
apparently doing something

```
cdef ndarray[double_t] foo

foo = np.empty(10,dtype='float64')

foo[0] = np.nan
```

actually assigns a python object (and does inference and such), so that you MUST have the gil,
but casting to a double seems to work though (as its now c-level)

ahojnnes commented:
You can reuse NPY_NAN cross-platform, e.g., https://github.com/scikit-image/scikit-image/blob/master/skimage/feature/_texture.pyx#L11

jreback commented:
@ahojnnes thxs, I did change it.

jreback commented:
I updated this. works for all groupbys now. So it gives a general 5-10% speedup on most operations (even single threaded), as there is slightly less generated code (in ``nogil`` it pretty much generates straight c code).

so we still need to hold the gil for the ``.resize`` operation, but its generally quite limited when this happens (e.g. resizing is a function of the number of uniques for example), but the cost of this is pretty low.

There IS a soln where I can not use numpy resize at ALL. But its quite a bit more c-code and IMHO not really worth it.

mrocklin commented:
Cool.  I'll try to run a few different groupby operations with dask.dataframe sometime early next week and report back.  I'm curious what were the operations that needed to be released?  Was there a core set of functionality or was it a large volume of trivial work?

jreback commented:
TL;DR.

So lots of 'boilerplate' changes in ``generated.pyx``, which is basically all of the algorithms for groupby ops themselves (e.g. sum,mean). These were straightforward insertions of ``with nogil``, just needed to avoid typing the ``object`` ones (which for the most part was easy; though a bit tricky on some non-groupby routines, e.g. see #10213).

However, the big speeds occur once I fixed the ``hashtable.pyx`` factorizer (e.g. the ``get_labels`` routine). The trivial soln actually produces a very odd perf issue (see [here](https://groups.google.com/forum/#!topic/cython-users/jtZ0FOGX2vU)), which is a bug / documentation issue.

In effect, you are using a ``nogil`` function that happens to have a ``with gil`` block WITHIN it, then the function is way slower than if you inline the exact same code.

I 'fixed' this issue by adding some additional c-structures to hold data (basically I changed the low-level implementation); these can be easily passed around; downside is that are a little 'messy' in the code department.

note to @shoyer I didn't actually bypass the realloc of memory (e.g. ``.resize``) of numpy arrays. though I do have a soln to that (basically manage a memory block myself, then turning that into a numpy array at the very end (in ``to_array``), easy enough, but was running into some compile issues, so dropped it in the end).

mrocklin commented:
I think that this is awesome and am afraid of it going stale.  Is there something blocking it from being merged into master?

shoyer commented:
@mrocklin We were waiting to get the 0.16.2 release out first, which happened last Friday. At this point I don't think there are any blockers.

jreback commented:
nothing blocking. I am going to update and see if I can hit some of #10213.

I have to do a bit of odd code-generation because for example you want to simultaneously support ``nogil`` (for numeric) and ``gil`` for ``object``, but its a context manager (and NOT a function, which has some odd per characteristics). IOW, you have to worry about formatting, a bit annoying.

jreback commented:
actually I do need to add a doc-note to the whatsnew (and maybe a small mention in the enhancingperf section as well).

mrocklin commented:
Hrm, yes, unfortunate that you can't manipulate the lock directly

jreback commented:
yep, would be nice to put in something along the lines of an ``#IFDEF``, but in reality the block is actually doing 'work', so maybe not easy/possible.

mrocklin commented:
Thought I'd caress this PR with a gentle ping.  

*ping*

jreback commented:
is that euphemism for harassment?

mrocklin commented:
Kind and loving harassment, yes.

jreback commented:
any commentary @jorisvandenbossche @shoyer @cpcloud @TomAugspurger 

going to have a followup for more GIL stuff. Its actually tricky making sure its doing (and you are measure exactly the right stuff)

cpcloud commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -56,8 +57,32 @@ Deprecations
+
+We are releasing the global-interpreter-lock (GIL) on some cython operations.
+This will allow other threads to run simultaneously during computation, potentially allowing performance improvements
+from multi-threading. Notably ``groupby`` and some indexing operations are a benefit to this. (:issue:`8882`)
```
i'd say "benefit from this"

cpcloud commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -56,8 +57,32 @@ Deprecations
+                   'data' : np.random.randn(N) })
+   df.groupby('key')['data'].sum()
+
+The releasing of the GIL could benefit an application that uses threads for user interactions (e.g. ``QT``), or performaning multi-threaded computations. A nice example of a library that can handle these types of computation-in-parallel is the dask_ library.
```
Releasing the GIL instead of "the releasing of the GIL"

cpcloud commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -56,8 +57,32 @@ Deprecations
+This will allow other threads to run simultaneously during computation, potentially allowing performance improvements
+from multi-threading. Notably ``groupby`` and some indexing operations are a benefit to this. (:issue:`8882`)
+
+For example the groupby expression in the following code will have the GIL released during the groups computation, e.g. ``df.groupby('key')``
```
maybe say "during the factorization step" instead of "during the groups computation", not totally clear what the "groups" computation is

cpcloud commented on pandas/hashtable.pyx:
```diff
@@ -80,72 +61,117 @@ cdef class ObjectVector:
+    int64_t *data
+    size_t n, m
+
+cdef uint8_t Int64VectorData_needs_resize(Int64VectorData *data) nogil:
```
I think you can used fused types to reduce the duplicity here.

```cython
ctypedef fused vector_data:
    Int64VectorData
    Float64VectorData

ctypedef fused sixty_four_bit_scalar:
    int64_t
    float64_t

cdef bint needs_resize(vector_data *data) nogil:
    return data.n == data.m

cdef void append_data(vector_data *data, sixty_four_bit_data x):
    data.data[data.n] = x
    data.n += 1
```

cpcloud commented on pandas/hashtable.pyx:
```diff
@@ -399,137 +324,166 @@ cdef class Int64HashTable: #(HashTable):
             raise KeyError(key)
 
-    def map(self, ndarray[int64_t] keys, ndarray[int64_t] values):
+    @cython.boundscheck(False)
```
do these also need `cython.wraparound(False)`?

jreback replied:
yes when you release the gil it gives a warning if you don't do this (as you are de-facto turning it off in c-land anyhow, I think it wants to make it explict)

cpcloud commented on pandas/hashtable.pyx:
```diff
@@ -740,7 +719,7 @@ cdef class PyObjectHashTable(HashTable):
             object val
             khiter_t k
-            ndarray[int64_t] locs = np.empty(n, dtype=np.int64)
+            int64_t[:] locs = np.empty(n, dtype=np.int64)
```
Was this change necessary or just for consistency with the rest of the HashTable implementations?

jreback replied:
just consistency

jreback commented on pandas/hashtable.pyx:
```diff
@@ -904,28 +864,34 @@ cdef class Int64Factorizer:
     cdef:
         Py_ssize_t i
         kh_int64_t *table
+        ndarray[int64_t] result_keys, result_counts
```
I think I tried to change this before and ran into an issue..trying again.

jreback commented on pandas/hashtable.pyx:
```diff
@@ -80,72 +61,120 @@ cdef class ObjectVector:
+    # compile time specilization of the fused types
+    # as the cross-product is generated, but we cannot assign float->int
+    # the types that don't pass are pruned
+    if (vector_data is Int64VectorData and sixty_four_bit_scalar is int64_t) or (
```
@cpcloud I got this to work after puzzling on what cython does with multiple fused types. It creates the cross-product, but you generally need only certain specilizations; it does this beaut as a compile-time prune, so no perf hit.

sinhrks commented:
Great job!!

mrocklin commented:
Woohoo!

*Pulls from master, checks benchmarks.*

jorisvandenbossche commented:
@mrocklin Can we already look forward to a new awesome blogpost? :-)

mrocklin commented:
Best I have to offer at the moment are the slowly growing [`dask.dataframe` docs](http://dask.pydata.org/en/latest/dataframe.html)

jorisvandenbossche commented:
I should definitely check it out once I find some time!

scari commented:
Wow! Great job! 👍🏻
